chacha20/src/cipher.rs

```diff
@@ -61,7 +62,7 @@ pub struct Cipher<R: Rounds> {
     buffer: Buffer,
 
     /// Position within buffer, or `None` if the buffer is not in use
-    buffer_pos: Option<u8>,
```

Comment:
IIRC this was a performance optimization for the AEAD use case, which is always able to work with ChaCha blocks-at-a-time until the last block.

It'd be good to double check how/if this impacts the performance of `chacha20poly1305`.

Reply:
The current API will not cache buffer if fed with data multiple of block size. Note the `if pos != 0 { .. }` and `if !rem.is_empty() { .. }` parts. The `Option<u8>` (which I think you've borrowed from the `ctr` crate) approach was redundant, since `buffer_pos` will never be equal to `Some(0)`.

Reply:
After quick and dirty tests (i.e. without fiddling with hardware settings) the `chacha20poly1305` performance is approximately the same on my CPU, with 2.605 cpb for decryption and 8.337 cpb for encryption (without enabled target-features).